### DISCUSSION TITLE
Status of the new Config

### DISCUSSION THREAD

**Discussion by hasezoey** [tramhao/termusic#302] [Show and tell]
re #301 (& #214)

Over the last week i have been working on the new config, i still have some work to-do (like translating from the old config and integrating the new into the rest of the application) but the new toml layout should be finish (from my view), so i wanted to gather some feedback.

Please note that the new config is *not* 1:1 (1 to 1) the same as the previous config, i took the liberty of adding some more user-choice to some places and changing some defaults (like lower default volume to not accidentally blow ears out).

I also have some questions that i would like resolved before posting a PR:
- ~~what is the purpose of [`playlist_display_symbol`](https://github.com/tramhao/termusic/blob/98498871d46ca5a9aa57a3cb8d30a9e2676d0a62/lib/src/config/mod.rs#L302)? it does not seem to be referenced anywhere (that is not the config itself or the config editor)~~
  After some more searching (and trying to remove it), i came to know that rust-analyzer just didnt show the references used in macros (like `format!()`), so it looked to me like it was unused (and had no proper documentation).
  TL;DR: this config option toggles whether "Loop Mode" will use text, or a symbol (ie `Mode: playlist` / `Mode: 🔁`)
- should extra key aliases be allowed to be parsed, if yes what aliases should there be? (like `esc` for `escape`)
  (some aliases i would add would be (alias, key) (`esc`, `escape`), (`space`, ` `), (`ctrl`, `control`), and maybe for the arrow keys without the `arrow` prefix)
- is serializing the keys as lowercase really a good idea? personally i find UPPERCASE (or maybe PascalCase) to be easier to read (example being `i` and `l` look *quite* similar in most common fonts, whereas `I` and `L` do not really look similar)
  (note that the parsing is case-insensitive)

new configs:

<details>
<summary>Server config default</summary>

```toml
version = "1"

[com]
port = 50101
address = "::"

[player]
music_dirs = ["/home/hasezoey/Music"]
library_scan_depth = 10
loop_mode = "playlist"
volume = 30
speed = 10
gapless = true
use_mediacontrols = true
set_discord_status = true

[player.remember_position]
music = "no"
podcast = "yes"

[player.seek_step]
short_tracks = 5
long_tracks = 30

[podcast]
concurrent_downloads_max = 3
max_download_retries = 3
download_dir = "/home/hasezoey/Music"
````

</details>

<details>
<summary>Server config advanced options</summary>

```toml
version = "1"

[com]
port = 50101
address = "::"

[player]
music_dirs = ["/home/hasezoey/Music"]
library_scan_depth = "unlimited"
# alternative
library_scan_depth = 10
loop_mode = "playlist"
volume = 30
speed = 10
gapless = true
use_mediacontrols = true
set_discord_status = true
remember_position = "yes"
seek_step = 10

# alternative
[player.remember_position]
music = 10
podcast = 30

# alternative
[player.remember_position]
music = "yes"
podcast = "no"

# alternative
[player.seek_step]
short_tracks = 5
long_tracks = 30

[podcast]
concurrent_downloads_max = 3
max_download_retries = 3
download_dir = "/home/hasezoey/Music"
```

</details>

<details>
<summary>TUI config default</summary>

```toml
version = "1"
com = "same"

[behavior]
quit_server_on_exit = true
confirm_quit = true

[coverart]
align = "bottom right"
size_scale = 0

[symbols]
library_highlight = "🦄"
playlist_highlight = "🚀"
playlist_current_track = "►"

[colors]
library_foreground = "Foreground"
library_background = "Reset"
library_border = "Blue"
library_highlight = "LightYellow"
playlist_foreground = "Foreground"
playlist_background = "Reset"
playlist_border = "Blue"
playlist_highlight = "LightYellow"
lyric_foreground = "Foreground"
lyric_background = "Reset"
lyric_border = "Blue"
progress_foreground = "LightBlack"
progress_background = "Reset"
progress_border = "Blue"
important_popup_foreground = "Red"
important_popup_background = "Reset"
important_popup_border = "Red"
generic_foreground = "Foreground"
generic_background = "Reset"
generic_border = "Blue"
generic_highlight = "LightYellow"

[theme]
name = "empty name"
author = "empty author"

[theme.primary]
background = "#000000"
foreground = "#ffffff"

[theme.cursor]
text = "#ffffff"
cursor = "#ffffff"

[theme.normal]
black = "#000000"
red = "#ff0000"
green = "#00ff00"
yellow = "#ffff00"
blue = "#0000ff"
magenta = "#ff00ff"
cyan = "#00ffff"
white = "#ffffff"

[theme.bright]
black = "#777777"
red = "#000000"
green = "#000000"
yellow = "#000000"
blue = "#000000"
magenta = "#000000"
cyan = "#000000"
white = "#000000"

[keys]
escape = "escape"
quit = "q"

[keys.view]
view_library = "1"
view_database = "2"
view_podcasts = "3"
open_config = "shift+c"
open_help = "control+h"

[keys.navigation]
up = "k"
down = "j"
left = "h"
right = "l"
goto_top = "g"
goto_bottom = "shift+g"

[keys.global_player]
toggle_pause = " "
next_track = "n"
previous_track = "shift+n"
volume_up = "+"
volume_down = "-"
seek_forward = "f"
seek_backward = "b"
speed_up = "control+f"
speed_down = "control+b"
toggle_prefetch = "control+g"
save_playlist = "control+s"

[keys.global_lyric]
lyric_adjust_forwards = "shift+f"
lyric_adjust_backwards = "shift+b"
lyric_cycle = "shift+t"

[keys.library]
load_dir = "shift+l"
delete = "d"
yank = "y"
paste = "p"
cycle_root = "o"
add_root = "a"
remove_root = "shift+a"
search = "/"
youtube_search = "s"
open_tag_editor = "t"

[keys.playlist]
delete = "d"
delete_all = "shift+d"
shuffle = "r"
cycle_loop_mode = "m"
play_selected = "l"
search = "/"
swap_up = "shift+k"
swap_down = "shift+j"
add_random_songs = "shift+s"
add_random_album = "s"

[keys.podcast]
search = "s"
mark_played = "m"
mark_all_played = "shift+m"
refresh_feed = "r"
refresh_all_feeds = "shift+r"
download_episode = "d"
delete_local_episode = "x"
delete_feed = "d"
delete_all_feeds = "shift+d"

[keys.adjust_cover_art]
move_left = "control+shift+arrowleft"
move_right = "control+shift+arrowright"
move_up = "control+shift+arrowup"
move_down = "control+shift+arrowdown"
increase_size = "control+shift+pageup"
decrease_size = "control+shift+pagedown"
toggle_hide = "control+shift+end"
```

</details>

the TUI config does not have as much "variability" as the server config in regards to alternatives (the only one currently being `com` can be `same` or a actual entry like in the server config, if we in the future could allow the tui to run on another system)

<details>
<summary>V1 config default as reference</summary>

```toml
music_dir = ["/home/hasezoey/Music", "/home/hasezoey/Music/mp3"]
player_port = 50101
player_interface = "::"
player_loop_mode = "Random"
player_volume = 70
player_speed = 10
player_gapless = true
podcast_simultanious_download = 3
podcast_max_retries = 3
podcast_dir = "/home/hasezoey/Music/podcast"
player_seek_step = "Auto"
player_remember_last_played_position = "Auto"
enable_exit_confirmation = true
playlist_display_symbol = true
playlist_select_random_track_quantity = 20
playlist_select_random_album_quantity = 5
theme_selected = "default"
kill_daemon_when_quit = true
player_use_mpris = true
player_use_discord = true

[album_photo_xywh]
x_between_1_100 = 100
y_between_1_100 = 77
width_between_1_100 = 20
align = "BottomRight"

[style_color_symbol]
library_foreground = "Foreground"
library_background = "Reset"
library_border = "Blue"
library_highlight = "LightYellow"
library_highlight_symbol = "🦄"
playlist_foreground = "Foreground"
playlist_background = "Reset"
playlist_border = "Blue"
playlist_highlight = "LightYellow"
playlist_highlight_symbol = "🚀"
progress_foreground = "LightBlack"
progress_background = "Reset"
progress_border = "Blue"
lyric_foreground = "Foreground"
lyric_background = "Reset"
lyric_border = "Blue"
currently_playing_track_symbol = "►"

[style_color_symbol.alacritty_theme]
path = ""
name = "default"
author = "Larry Hao"
background = "#101421"
foreground = "#fffbf6"
cursor = "#FFFFFF"
text = "#1E1E1E"
black = "#2e2e2e"
red = "#eb4129"
green = "#abe047"
yellow = "#f6c744"
blue = "#47a0f3"
magenta = "#7b5cb0"
cyan = "#64dbed"
white = "#e5e9f0"
light_black = "#565656"
light_red = "#ec5357"
light_green = "#c0e17d"
light_yellow = "#f9da6a"
light_blue = "#49a4f8"
light_magenta = "#a47de9"
light_cyan = "#99faf2"
light_white = "#ffffff"

[keys.global_esc.code]
type = "Esc"

[keys.global_esc.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_quit.code]
type = "Char"
args = "q"

[keys.global_quit.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_left.code]
type = "Char"
args = "h"

[keys.global_left.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_down.code]
type = "Char"
args = "j"

[keys.global_down.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_up.code]
type = "Char"
args = "k"

[keys.global_up.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_right.code]
type = "Char"
args = "l"

[keys.global_right.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_goto_top.code]
type = "Char"
args = "g"

[keys.global_goto_top.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_goto_bottom.code]
type = "Char"
args = "G"

[keys.global_goto_bottom.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_player_toggle_pause.code]
type = "Char"
args = " "

[keys.global_player_toggle_pause.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_player_next.code]
type = "Char"
args = "n"

[keys.global_player_next.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_player_previous.code]
type = "Char"
args = "N"

[keys.global_player_previous.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_player_volume_plus_1.code]
type = "Char"
args = "+"

[keys.global_player_volume_plus_1.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_player_volume_plus_2.code]
type = "Char"
args = "="

[keys.global_player_volume_plus_2.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_player_volume_minus_1.code]
type = "Char"
args = "_"

[keys.global_player_volume_minus_1.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_player_volume_minus_2.code]
type = "Char"
args = "-"

[keys.global_player_volume_minus_2.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_help.code]
type = "Char"
args = "h"

[keys.global_help.modifier]
type = "KeyModifiers"
bits = 2

[keys.global_player_seek_forward.code]
type = "Char"
args = "f"

[keys.global_player_seek_forward.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_player_seek_backward.code]
type = "Char"
args = "b"

[keys.global_player_seek_backward.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_lyric_adjust_forward.code]
type = "Char"
args = "F"

[keys.global_lyric_adjust_forward.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_lyric_adjust_backward.code]
type = "Char"
args = "B"

[keys.global_lyric_adjust_backward.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_player_speed_up.code]
type = "Char"
args = "f"

[keys.global_player_speed_up.modifier]
type = "KeyModifiers"
bits = 2

[keys.global_player_speed_down.code]
type = "Char"
args = "b"

[keys.global_player_speed_down.modifier]
type = "KeyModifiers"
bits = 2

[keys.global_lyric_cycle.code]
type = "Char"
args = "T"

[keys.global_lyric_cycle.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_layout_treeview.code]
type = "Char"
args = "1"

[keys.global_layout_treeview.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_layout_database.code]
type = "Char"
args = "2"

[keys.global_layout_database.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_player_toggle_gapless.code]
type = "Char"
args = "g"

[keys.global_player_toggle_gapless.modifier]
type = "KeyModifiers"
bits = 2

[keys.global_config_open.code]
type = "Char"
args = "C"

[keys.global_config_open.modifier]
type = "KeyModifiers"
bits = 1

[keys.global_save_playlist.code]
type = "Char"
args = "s"

[keys.global_save_playlist.modifier]
type = "KeyModifiers"
bits = 2

[keys.global_layout_podcast.code]
type = "Char"
args = "3"

[keys.global_layout_podcast.modifier]
type = "KeyModifiers"
bits = 0

[keys.global_xywh_move_left.code]
type = "Left"

[keys.global_xywh_move_left.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_move_right.code]
type = "Right"

[keys.global_xywh_move_right.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_move_up.code]
type = "Up"

[keys.global_xywh_move_up.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_move_down.code]
type = "Down"

[keys.global_xywh_move_down.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_zoom_in.code]
type = "PageUp"

[keys.global_xywh_zoom_in.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_zoom_out.code]
type = "PageDown"

[keys.global_xywh_zoom_out.modifier]
type = "KeyModifiers"
bits = 3

[keys.global_xywh_hide.code]
type = "End"

[keys.global_xywh_hide.modifier]
type = "KeyModifiers"
bits = 3

[keys.library_load_dir.code]
type = "Char"
args = "L"

[keys.library_load_dir.modifier]
type = "KeyModifiers"
bits = 1

[keys.library_delete.code]
type = "Char"
args = "d"

[keys.library_delete.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_yank.code]
type = "Char"
args = "y"

[keys.library_yank.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_paste.code]
type = "Char"
args = "p"

[keys.library_paste.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_search.code]
type = "Char"
args = "/"

[keys.library_search.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_search_youtube.code]
type = "Char"
args = "s"

[keys.library_search_youtube.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_tag_editor_open.code]
type = "Char"
args = "t"

[keys.library_tag_editor_open.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_switch_root.code]
type = "Char"
args = "o"

[keys.library_switch_root.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_add_root.code]
type = "Char"
args = "a"

[keys.library_add_root.modifier]
type = "KeyModifiers"
bits = 0

[keys.library_remove_root.code]
type = "Char"
args = "A"

[keys.library_remove_root.modifier]
type = "KeyModifiers"
bits = 1

[keys.playlist_delete.code]
type = "Char"
args = "d"

[keys.playlist_delete.modifier]
type = "KeyModifiers"
bits = 0

[keys.playlist_delete_all.code]
type = "Char"
args = "D"

[keys.playlist_delete_all.modifier]
type = "KeyModifiers"
bits = 1

[keys.playlist_shuffle.code]
type = "Char"
args = "r"

[keys.playlist_shuffle.modifier]
type = "KeyModifiers"
bits = 0

[keys.playlist_mode_cycle.code]
type = "Char"
args = "m"

[keys.playlist_mode_cycle.modifier]
type = "KeyModifiers"
bits = 0

[keys.playlist_play_selected.code]
type = "Char"
args = "l"

[keys.playlist_play_selected.modifier]
type = "KeyModifiers"
bits = 0

[keys.playlist_search.code]
type = "Char"
args = "/"

[keys.playlist_search.modifier]
type = "KeyModifiers"
bits = 0

[keys.playlist_swap_down.code]
type = "Char"
args = "J"

[keys.playlist_swap_down.modifier]
type = "KeyModifiers"
bits = 1

[keys.playlist_swap_up.code]
type = "Char"
args = "K"

[keys.playlist_swap_up.modifier]
type = "KeyModifiers"
bits = 1

[keys.playlist_cmus_lqueue.code]
type = "Char"
args = "S"

[keys.playlist_cmus_lqueue.modifier]
type = "KeyModifiers"
bits = 1

[keys.playlist_cmus_tqueue.code]
type = "Char"
args = "s"

[keys.playlist_cmus_tqueue.modifier]
type = "KeyModifiers"
bits = 0

[keys.database_add_all.code]
type = "Char"
args = "L"

[keys.database_add_all.modifier]
type = "KeyModifiers"
bits = 1

[keys.config_save.code]
type = "Char"
args = "s"

[keys.config_save.modifier]
type = "KeyModifiers"
bits = 2

[keys.podcast_mark_played.code]
type = "Char"
args = "m"

[keys.podcast_mark_played.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_mark_all_played.code]
type = "Char"
args = "M"

[keys.podcast_mark_all_played.modifier]
type = "KeyModifiers"
bits = 1

[keys.podcast_episode_download.code]
type = "Char"
args = "d"

[keys.podcast_episode_download.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_episode_delete_file.code]
type = "Char"
args = "x"

[keys.podcast_episode_delete_file.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_delete_feed.code]
type = "Char"
args = "d"

[keys.podcast_delete_feed.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_delete_all_feeds.code]
type = "Char"
args = "D"

[keys.podcast_delete_all_feeds.modifier]
type = "KeyModifiers"
bits = 1

[keys.podcast_search_add_feed.code]
type = "Char"
args = "s"

[keys.podcast_search_add_feed.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_refresh_feed.code]
type = "Char"
args = "r"

[keys.podcast_refresh_feed.modifier]
type = "KeyModifiers"
bits = 0

[keys.podcast_refresh_all_feeds.code]
type = "Char"
args = "R"

[keys.podcast_refresh_all_feeds.modifier]
type = "KeyModifiers"
bits = 1
```

</details>

**Comment by hasezoey** [comment]
@tramhao @MyyPo (because we are on the topic already, #314), what would be better for the `colors` field, use it as it is currently in the examples provided, or should they be "outsourced" / sub-classed like the `keys` fields?

**Comment by myypo** [comment]
@hasezoey  I can't state any strong preferences since I am not that knowledgeble in toml conventions, but in other languages I would have nested stuff. Like having `colors.playlist` and such. Maybe later there will be more customization fields, in this case it might be better to have `style.playlist` kind of fields with `_color` subfields. It might make sense to have these definitions closer to each other as they grow without introducing 3-level nesting. 

On a separate note: `generic_*` fields might be too generic :D as in case of the UI growing there might end up appearing a temptation to throw too much stuff here limiting the customization.

**Reply by hasezoey** [reply]
> On a separate note: generic_* fields might be too generic :D as in case of the UI growing there might end up appearing a temptation to throw too much stuff here limiting the customization.

it was written with the intend on the current termusic state and the near-future state, where basically `library_*` fields were used as generic fields.
i guess i could rename it to `fallback_*` that way it would also be clear and clear in the future if more customization is added.

> Maybe later there will be more customization fields, in this case it might be better to have style.playlist kind of fields with _color subfields.

could you explain more what you mean with `style.` in this context? also maybe a example of how it would look?

**Reply by myypo** [reply]
> it was written with the intend on the current termusic state and the near-future state, where basically `library_*` fields were used as generic fields.
> i guess i could rename it to `fallback_*` that way it would also be clear and clear in the future if more customization is added.

Then `fallback_*` is a great name, I think. 

> could you explain more what you mean with `style.` in this context? also maybe a example of how it would look?

Something like this.

```toml
[style.playlist]
foreground_color = "Foreground"
background_color = "Reset"
border = "Rounded" # New thing
border_color = "Blue"
highlight_color = "LightYellow"
```

**Reply by hasezoey** [reply]
> Something like this.

good call, will consider; then should i also merge the `symbols` back into that `style` too?

**Reply by myypo** [reply]
Seems to make sense if we consider locality to be a virtue. Should be a bit simpler for a user to reason about it this way.

**Reply by hasezoey** [reply]
implemented, now it looks like (snippit, not whole config):

```toml
[style.library]
foreground_color = "Foreground"
background_color = "Reset"
border_color = "Blue"
highlight_color = "LightYellow"
highlight_symbol = "🦄"

[style.playlist]
foreground_color = "Foreground"
background_color = "Reset"
border_color = "Blue"
highlight_color = "LightYellow"
highlight_symbol = "🚀"
current_track_symbol = "►"
use_loop_mode_symbol = true

[style.lyric]
foreground_color = "Foreground"
background_color = "Reset"
border_color = "Blue"

[style.progress]
foreground_color = "LightBlack"
background_color = "Reset"
border_color = "Blue"

[style.important_popup]
foreground_color = "Yellow"
background_color = "Reset"
border_color = "Yellow"

[style.fallback]
foreground_color = "Foreground"
background_color = "Reset"
border_color = "Blue"
highlight_color = "LightYellow"
```
EDIT: forgot to rename the fields

instead of the `[colors]` and `[symbols]` sections

@tramhao what do you think, is the old v1, middle v2 style or the new (in this comment) style the best?
